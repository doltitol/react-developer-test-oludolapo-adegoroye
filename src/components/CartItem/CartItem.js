import React, { PureComponent } from 'react';
import './cartitem.style.scss';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import { addQuantity, removeQuantity } from '../../redux/actions/cartActions';
import AttributeButton from '../Buttons/AttributeButton';
import ColorButton from '../Buttons/ColorButton';
import CountButton from '../Buttons/CountButton';
import CartImageSlider from '../CartImageSlider/CartImageSlider';
import { numberCommaFormatter } from '../../util';

export class CartItem extends PureComponent {
  render() {
    const pricing = this.props.cartItem.prices.filter(
      (price) => price.currency.symbol === this.props.currency
    );
    const color = this.props.cartItem.attributes.filter(
      (attr) => attr.id === 'Color'
    )
      ? this.props.cartItem.attributes.filter((attr) => attr.id === 'Color')
      : null;
    const size = this.props.cartItem.attributes.filter(
      (attr) => attr.id === 'Size' || attr.id === 'Capacity'
    )
      ? this.props.cartItem.attributes.filter((attr) => attr.id !== 'Color')
      : null;
    return (
      <div className='cart-item' data-testid='cart-item'>
        <div className='cart-item-details'>
          <Link
            to={`/product/${this.props.cartItem.id}`}
            onClick={() => {
              this.props.closeCart();
            }}
          >
            <p className='cart-item-details-brand'>
              {this.props.cartItem.brand}
            </p>
          </Link>
          <Link
            to={`/product/${this.props.cartItem.id}`}
            onClick={() => {
              this.props.closeCart();
            }}
          >
            <p className='cart-item-details-name'>{this.props.cartItem.name}</p>
          </Link>
          <p className='cart-item-details-pricing'>
            {pricing[0].currency.symbol}{' '}
            {numberCommaFormatter(pricing[0].amount)}
          </p>
          {size.length !== 0 && (
            <div className='cart-item-details-attribute'>
              <p className='cart-item-details-attribute-text'>
                {size[0].name}:
              </p>
              <div className='cart-item-details-attribute-items'>
                {size[0].items.map((attribute) => (
                  <div
                    key={attribute.id}
                    className='cart-item-details-attribute-items-item'
                  >
                    <AttributeButton
                      text={attribute.value}
                      active={
                        this.props.cartItem.activeAttribute === attribute.id
                      }
                    />
                  </div>
                ))}
              </div>
            </div>
          )}
          {color.length !== 0 && (
            <div className='cart-item-details-attribute'>
              <p className='cart-item-details-attribute-text'>
                {color[0].name}:
              </p>
              <div className='cart-item-details-attribute-items'>
                {color[0].items.map((color) => (
                  <div
                    key={color.id}
                    className='cart-item-details-attribute-items-item'
                  >
                    <ColorButton
                      color={color.value}
                      active={this.props.cartItem.activeColor === color.id}
                    />
                  </div>
                ))}
              </div>
            </div>
          )}
        </div>

        <div className='cart-item-buttons'>
          <CountButton
            text='+'
            onClick={() => this.props.addQuantity(this.props.cartItem)}
          />
          <p className='cart-item-buttons-text'>{this.props.cartItem.qty}</p>
          <CountButton
            text='-'
            onClick={() => this.props.removeQuantity(this.props.cartItem)}
          />
        </div>
        <div className='cart-item-Image'>
          <CartImageSlider gallery={this.props.cartItem.gallery} />
        </div>
      </div>
    );
  }
}
const mapStateToProps = (state) => {
  return {
    cartItems: state.cart.cartItems,
    currency: state.cart.currency,
    cartQuantity: state.cart.cartQuantity,
    total: state.cart.total,
  };
};
const mapDispatchToProps = (dispatch) => {
  return {
    addQuantity: (product) => {
      dispatch(addQuantity(product));
    },
    removeQuantity: (product) => {
      dispatch(removeQuantity(product));
    },
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(CartItem);
