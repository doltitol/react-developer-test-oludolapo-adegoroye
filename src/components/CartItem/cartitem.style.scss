@import '../../styles/index.scss';

.cart-item {
  border-bottom: 1px solid $border;
  display: flex;
  flex-direction: row;
  justify-content: space-between;
  align-items: center;
  padding: 20px 0;
  @media (max-width: 600px){
     display: flex;
     flex-wrap: wrap;
  }
  &-details {
    width: 85%;
    display: flex;
    flex-direction: column;
    justify-content: flex-start;
    height: 288px;
     @media (max-width: 600px){
  width: 100%;
  }
    &-brand {
      @include raleway(24px, 27px, 600);
    }
    &-name {
      @include raleway(24px, 27px, 400);
      margin-top: 5px;
    }
    &-pricing {
      @include raleway(20px, 24px, 700);
      margin-top: 5px;
    }
    &-attribute {
      margin-top: 20px;
      &-text {
        text-transform: uppercase;
        @include robotoCondensed(18px, 18px, 700);
      }
      &-items{
      margin-top: 5px;
      display: flex;
      flex-direction: row;
      &-item{
      margin-right: 10px;
      }
      }
    }
  }
  &-buttons {
  width: 10%;
  height: 288px;
display: flex;
flex-direction: column;
justify-content: space-between;
align-items: center;
&-text{
@include raleway(24px, 25.6px, 500);
}
  }
  &-image {
    width: 20%;
  }
}
