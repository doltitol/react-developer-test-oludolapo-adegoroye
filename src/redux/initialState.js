export const initialState = {
  cartItems: JSON.parse(localStorage.getItem('cartItems') || '[]'),
  cartQuantity: JSON.parse(localStorage.getItem('cartItems') || '[]').reduce(
    (acc, item) => acc + item.qty,
    0
  ),
  currency: '$',
};
