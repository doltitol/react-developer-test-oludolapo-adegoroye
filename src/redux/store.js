import { createStore, combineReducers } from 'redux';
import cartReducer from './reducers/cartReducers';

const rootReducer = combineReducers({
  cart: cartReducer,
});
export const store = createStore(rootReducer);
