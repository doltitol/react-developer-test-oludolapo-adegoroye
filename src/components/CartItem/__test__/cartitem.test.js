import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import { BrowserRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import { store } from '../../../redux/store';
import CartItem from '../CartItem';

describe('<CartItem /> spec', () => {
  it('renders the empty component', () => {
    const props = {
      cartItem: {
        __typename: 'Product',
        id: 'apple-iphone-12-pro',
        name: 'iPhone 12 Pro',
        inStock: true,
        gallery: [
          'https://store.storeimages.cdn-apple.com/4982/as-images.apple.com/is/iphone-12-pro-family-hero?wid=940&amp;hei=1112&amp;fmt=jpeg&amp;qlt=80&amp;.v=1604021663000',
        ],
        description: 'This is iPhone 12. Nothing else to say.',
        category: 'tech',
        attributes: [
          {
            __typename: 'AttributeSet',
            id: 'Capacity',
            name: 'Capacity',
            type: 'text',
            items: [
              {
                __typename: 'Attribute',
                displayValue: '512G',
                value: '512G',
                id: '512G',
              },
              {
                __typename: 'Attribute',
                displayValue: '1T',
                value: '1T',
                id: '1T',
              },
            ],
          },
          {
            __typename: 'AttributeSet',
            id: 'Color',
            name: 'Color',
            type: 'swatch',
            items: [
              {
                __typename: 'Attribute',
                displayValue: 'Green',
                value: '#44FF03',
                id: 'Green',
              },
              {
                __typename: 'Attribute',
                displayValue: 'Cyan',
                value: '#03FFF7',
                id: 'Cyan',
              },
              {
                __typename: 'Attribute',
                displayValue: 'Blue',
                value: '#030BFF',
                id: 'Blue',
              },
              {
                __typename: 'Attribute',
                displayValue: 'Black',
                value: '#000000',
                id: 'Black',
              },
              {
                __typename: 'Attribute',
                displayValue: 'White',
                value: '#FFFFFF',
                id: 'White',
              },
            ],
          },
        ],
        prices: [
          {
            __typename: 'Price',
            currency: {
              __typename: 'Currency',
              label: 'USD',
              symbol: '$',
            },
            amount: 1000.76,
          },
          {
            __typename: 'Price',
            currency: {
              __typename: 'Currency',
              label: 'GBP',
              symbol: '£',
            },
            amount: 719.34,
          },
          {
            __typename: 'Price',
            currency: {
              __typename: 'Currency',
              label: 'AUD',
              symbol: 'A$',
            },
            amount: 1290.99,
          },
          {
            __typename: 'Price',
            currency: {
              __typename: 'Currency',
              label: 'JPY',
              symbol: '¥',
            },
            amount: 108074.6,
          },
          {
            __typename: 'Price',
            currency: {
              __typename: 'Currency',
              label: 'RUB',
              symbol: '₽',
            },
            amount: 75680.48,
          },
        ],
        brand: 'Apple',
        activeColor: 'Green',
        activeAttribute: '512G',
        qty: 1,
        cartId: 2,
      },
    };
    const container = render(
      <Provider store={store}>
        <BrowserRouter>
          <CartItem {...props} />
        </BrowserRouter>
      </Provider>
    );
    const getByTestId = container.getByTestId;
    const cartItem = getByTestId('cart-item');
    expect(cartItem.children).toHaveLength(3);
    expect(container.children).toMatchSnapshot();
  });
});
