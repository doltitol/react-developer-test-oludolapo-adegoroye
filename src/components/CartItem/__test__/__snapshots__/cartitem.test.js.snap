// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartItem /> spec renders the empty component 1`] = `undefined`;
